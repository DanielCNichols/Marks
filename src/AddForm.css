* {
  box-sizing: border-box;
}

.add-form fieldset {
  border: none;
  width: 100%;
}

.add-form legend {
  color: #68cffa;
  font-size: 1.3em;
  font-weight: 800;
}

.add-form label {
  display: block;
  margin: 10px 0;
  color: #23a840;
  font-weight: 600;
}
.add-form input[type='text'],
select {
  display: block;
  margin: 10px auto;
  border-radius: 5px;
  height: 2em;
  padding: 5px;
}

.add-form select {
  width: 175px;
  margin: 10px auto;
}

.add-form button {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 10px auto;
  padding: 5px;
  text-align: center;
  font-family: inherit;
}

.add-form button > span {
  color: rgba(255, 255, 255, 0.8);
  font-size: 1rem;
  display: block;
  width: auto;
  margin: 0 auto;
}

.add-form-desc > textarea {
  width: 175px;
  display: block;
  margin: 10px auto;
}

.add-form-controls {
  display: flex;
  justify-content: space-around;
  margin-bottom: 10px;
}

.add-form-controls button {
  background: transparent;
  border: none;
  margin: 5px 0;
  font-size: 2em;
}

@media screen and (min-width: 800px) {
  .add-form {
    width: 100%;
    margin: 0 auto;
    text-align: center;
  }

  .add-form-controls {
    width: 70%;
    margin: 0 auto;
    display: flex;
    flex-direction: row;
    align-items: space-between;
  }
}
