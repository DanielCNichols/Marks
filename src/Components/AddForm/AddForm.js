import React, { useState } from 'react';
import ApiService from '../../Services/api-service';
import s from './AddForm.module.css';
import { MdCancel, MdSave } from 'react-icons/md';
import isUrl from 'isurl';

export default function AddForm(props) {
  const [error, setError] = useState(null); // Do I need to make another piece of state here? Could there be a"global error" in this case?

  async function handleSubmit(e) {
    try {
      e.preventDefault();
      const { title, url, desc, rating } = e.target;

      if (!title.value || !url.value) {
        throw new Error('Title and Url are required');
      }

      if (!isUrl(new URL(url.value))) {
        throw new Error('Must be a valid url');
      }

      const newBookmark = {
        title: title.value,
        url: url.value,
        desc: desc.value,
        rating: rating.value,
      };

      let res = await ApiService.postBookmark(newBookmark);
      props.addBookmark(res);
      props.addToggle();
    } catch (error) {
      setError(error);
    }
  }

  const { addToggle } = props;
  return (
    <form className={s.addForm} onSubmit={handleSubmit}>
      <fieldset>
        <legend>Add Bookmark</legend>
        <div className={s.formElement}>
          <label>Title</label>
          <input
            type="text"
            name="title"
            aria-required="true"
            aria-label="Title"
          />
        </div>
        <div className={s.formElement}>
          <label>Url</label>
          <input type="text" name="url" aria-required="true" aria-label="Url" />
        </div>
        <div className={s.formElement}>
          <label>Rating</label>
          <select
            name="rating"
            aria-required="false"
            aria-label="Select Rating"
          >
            <option value="1">1</option>
            <option value="2">2</option>
            <option value="3">3</option>
            <option value="4">4</option>
            <option value="5">5</option>
          </select>
        </div>
        <div className={s.formElement}>
          <label>Description</label>
          <textarea
            name="desc"
            rows="5"
            id="desc"
            aria-required="false"
            aria-label="Description"
          />
        </div>
        {error && (
          <>
            <p style={{ color: 'white' }}>{error.message}</p>
          </>
        )}
        <div className={s.formControls}>
          <button className={s.cancel} type="button" onClick={addToggle}>
            <MdCancel />
            <p>Cancel</p>
          </button>
          <button className={s.submit} type="submit">
            <MdSave />
            <p>Save</p>
          </button>
        </div>
      </fieldset>
    </form>
  );
}
