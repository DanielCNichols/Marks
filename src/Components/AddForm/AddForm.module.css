@import '../../vars.module.css';
* {
  box-sizing: border-box;
}

.error,
.postError {
  color: red;
  text-align: center;
  font-size: 0.8em;
}

.postError {
  margin: 0 auto 10px auto;
  width: 90%;
}

.addForm {
  margin: 0 auto;
  width: 100%;
  border: 1px solid rgba(255, 255, 255, 0.38);
  border-radius: 6px;
  padding: 20px 10px;
}

.addForm fieldset,
.formElement {
  display: flex;
  flex-direction: column;
  width: 90%;
  margin: auto;
}

.addForm legend {
  text-align: center;
  color: var(--color-subheading);
  margin-bottom: 20px;
  font-weight: bold;
}

.formElement {
  margin-bottom: 10px;
}

.formElement input,
.formElement select {
  border-radius: 3px;
  height: 30px;
  padding: 3px 5px;
}

.formElement label {
  margin-bottom: 5px;
}

.formControls {
  display: flex;
  justify-content: space-around;
  align-items: space-around;
  margin-top: 20px;
}

.formControls button {
  border: none;
  background-color: transparent;
  outline: none;
}

.submit svg {
  color: var(--color-secondary);
  font-size: 3em;
}

.cancel svg {
  color: red;
  font-size: 3em;
}

.submit,
.cancel {
  color: white;
}
/* tablet styles */

/* Desktop Styles */

@media screen and (min-width: 800px) {
  .addForm {
    width: 400px;
  }
}
