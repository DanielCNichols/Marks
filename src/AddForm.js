import React from 'react';
import ApiService from './api-service';
import './AddForm.css';
import { MdCancel, MdSave } from 'react-icons/md';

export default class AddForm extends React.Component {
  state = {
    error: null,
  };

  //Try to make this a cool popout.

  handleSubmit = e => {
    e.preventDefault();
    const { title, url, desc, rating } = e.target;
    const newBookmark = {
      title: title.value,
      url: url.value,
      desc: desc.value,
      rating: rating.value,
    };

    ApiService.postBookmark(newBookmark)
      .then(res => {
        this.props.addBookmark(res);
        this.props.addToggle();
      })
      .catch(error => this.setState({ error: error }));
  };

  renderError() {
    const { error } = this.state;
    return (
      <>
        {error ? (
          <div className="error" aria-live="assertive">
            {error.message}
          </div>
        ) : null}
      </>
    );
  }

  render() {
    const { addToggle } = this.props;
    return (
      <form className="add-form" onSubmit={this.handleSubmit}>
        <fieldset>
          <legend>Add Bookmark</legend>
          <label className="add-form-title">
            Title
            <input
              type="text"
              name="title"
              aria-required="true"
              aria-label="Title"
            />
          </label>
          <label className="add-form-url">
            Url
            <input
              type="text"
              name="url"
              aria-required="true"
              aria-label="Url"
            />
          </label>
          <label className="add-form-rating">
            Rating
            <select
              name="rating"
              aria-required="false"
              aria-label="Select Rating"
            >
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
              <option value="4">4</option>
              <option value="5">5</option>
            </select>
          </label>
          <label className="add-form-desc">
            Description
            <textarea
              name="desc"
              rows="5"
              id="desc"
              aria-required="false"
              aria-label="Description"
            />
          </label>
          {this.renderError()}
          <div className="add-form-controls">
            <button className="cancel" type="button" onClick={addToggle}>
              <MdCancel />
              <span>Cancel</span>
            </button>
            <button className="save" type="submit">
              <MdSave />
              <span>Save</span>
            </button>
          </div>
        </fieldset>
      </form>
    );
  }
}
