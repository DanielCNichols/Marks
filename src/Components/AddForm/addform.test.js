import React from 'react';
import ReactDOM from 'react-dom';
import AddForm from './AddForm';
import renderer from 'react-test-renderer';

describe('Add form tests', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AddForm />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer.create(<AddForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
