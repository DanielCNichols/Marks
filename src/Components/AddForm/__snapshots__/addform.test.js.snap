// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add form tests renders the UI as expected 1`] = `
<form
  className="addForm"
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      Add Bookmark
    </legend>
    <div
      className="formElement"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        aria-label="Title"
        aria-placeholder="Reddit"
        aria-required="true"
        id="title"
        name="title"
        placeholder="Reddit"
        type="text"
      />
    </div>
    <div
      className="formElement"
    >
      <label
        htmlFor="url"
      >
        Url
      </label>
      <input
        aria-label="Url"
        aria-placehder="https://www.reddit.com"
        aria-required="true"
        id="url"
        name="url"
        placeholder="https://www.reddit.com"
        type="text"
      />
    </div>
    <div
      className="formElement"
    >
      <label
        htmlFor="rating"
      >
        Rating
      </label>
      <select
        aria-label="Select Rating"
        aria-required="false"
        id="rating"
        name="rating"
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="5"
        >
          5
        </option>
      </select>
    </div>
    <div
      className="formElement"
    >
      <label
        htmlFor="description"
      >
        Description
      </label>
      <textarea
        aria-label="Description"
        aria-placeholder="This is the best website ever."
        aria-required="false"
        id="description"
        name="desc"
        placeholder="This is the best website ever."
        rows="5"
      />
    </div>
    <div
      className="formControls"
    >
      <button
        className="cancel"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
        <p>
          Cancel
        </p>
      </button>
      <button
        className="submit"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
          />
        </svg>
        <p>
          Save
        </p>
      </button>
    </div>
  </fieldset>
</form>
`;
